File: AoC2015/Day1/aoc2015_day1_part_a.js
/**
 * Program to calculate which floor the instructions 
 * from the text file will take Santa to in the apartment 
 * building. 
 */

/** Functions */
function calculateSantaFloor(instructions) {
    /**
     * Function to calculate which floor Santa ends up
     * on after following the instructions in the text 
     * file.
     */
    // Santa starts at floor zero
    var floorNumber = 0;
    // Iterate over instructions
    for (instruction of instructions) {
        if (instruction === '(') {
            floorNumber++;
        } else if (instruction === ')') {
            floorNumber--;
        }
    }
    // Return the floor number
    return Math.abs(floorNumber);
}

/** Driver Program */
// Santa's instructions
instructions = null;
// Require the filesystem module
const fs = require('fs');
// Require the assertions module
const assert = require("assert");
// Read contents of file as a string synchronously
try {
    instructions = fs.readFileSync('aoc2015_day1_input.txt', 'utf8');
} catch (err) {
    if (err) console.error(err);
}
// Some assertions to check the function is working correctly
try {
    // Sample I/O assertions
    assert.equal(calculateSantaFloor("(())"), 0, "TEST1A");
    assert.equal(calculateSantaFloor("(())"), calculateSantaFloor("()()"), "TEST1B");
    assert.equal(calculateSantaFloor("((("), 3, "TEST2A");
    assert.equal(calculateSantaFloor("((("), calculateSantaFloor("(()(()("), "TEST2B");
    assert.equal(calculateSantaFloor("))((((("), 3, "TEST3");
    assert.equal(calculateSantaFloor("())"), 1, "TEST4A");
    assert.equal(calculateSantaFloor("())"), calculateSantaFloor("))("), "TEST4B");
    assert.equal(calculateSantaFloor(")))"), 3, "TEST5A");
    assert.equal(calculateSantaFloor(")))"), calculateSantaFloor(")())())"), "TEST5B");
    // Final assertion on challenge input
    assert.equal(calculateSantaFloor(instructions), 232, "TESTN");
} catch (err) {
    console.error(err);
}

File: AoC2015/Day1/aoc2015_day1_part_a.py
"""
Program to calculate which floor the instructions 
from the text file will take Santa to in the apartment 
building. 
"""
def calculate_santa_floor(instructions):
    """
    Function to calculate which floor Santa ends up
    on after following the instructions in the text 
    file.
    """
    # Santa starts at floor zero
    floor_number = 0
    # Iterate over instructions
    for instruction in instructions:
        if instruction == '(':
            floor_number += 1
        elif instruction == ')':
            floor_number -= 1
    return abs(floor_number)

if __name__ == '__main__':
    # This variable holds the instructions for Santa
    instructions = None
    # Open the file for processing
    with open("aoc2015_day1_input.txt", 'r') as f:
        # Reads contents of file as a single string
        instructions = f.read()
        # Close the file after processing
        f.close()
    # Some assertions to check the function is working correctly
    assert calculate_santa_floor("(())") == calculate_santa_floor("()()") == 0
    assert calculate_santa_floor("(((") == calculate_santa_floor("(()(()(") == 3
    assert calculate_santa_floor("))(((((") == 3
    assert calculate_santa_floor("())") == calculate_santa_floor("))(") == 1
    assert calculate_santa_floor(")))") == calculate_santa_floor(")())())") == 3
    # Print the solution to the challenge with the text file input
    print(calculate_santa_floor(instructions))
